x-pack/plugin/ccr/src/test/java/org/elasticsearch/xpack/ccr/repository/CcrRestoreSourceServiceTests.java

```diff
@@ -122,4 +127,51 @@ public void testCloseShardListenerFunctionality() throws IOException {
         restoreSourceService.closeSession(sessionUUID3);
         closeShards(indexShard1, indexShard2);
     }
+
+    public void testGetSessionReader() throws IOException {
+        IndexShard indexShard1 = newStartedShard(true);
+        final String sessionUUID1 = UUIDs.randomBase64UUID();
+
+        restoreSourceService.openSession(sessionUUID1, indexShard1);
+
+        ArrayList<StoreFileMetaData> files = new ArrayList<>();
+        indexShard1.snapshotStoreMetadata().forEach(files::add);
+
+        StoreFileMetaData fileMetaData = files.get(0);
+        String fileName = fileMetaData.name();
+
+        byte[] expectedBytes = new byte[(int) fileMetaData.length()];
+        byte[] actualBytes = new byte[(int) fileMetaData.length()];
+        Engine.IndexCommitRef indexCommitRef = indexShard1.acquireSafeIndexCommit();
+        try (IndexInput indexInput = indexCommitRef.getIndexCommit().getDirectory().openInput(fileName, IOContext.READONCE)) {
+            indexInput.seek(0);
+            indexInput.readBytes(expectedBytes, 0, (int) fileMetaData.length());
+        }
+
+        try (CcrRestoreSourceService.Reader reader = restoreSourceService.getSessionReader(sessionUUID1, fileName)) {
```

Comment:
also test that switching between files does not leak?

Reply:
Changed.